vowpalwabbit/parse_args.cc

```diff
@@ -1230,17 +1230,20 @@ void register_reductions(vw& all, std::vector<reduction_setup_fn>& reductions)
 
   for (auto setup_fn : reductions)
   {
-    if (allowlist.count(setup_fn)) { all.reduction_stack.push(std::make_tuple(allowlist[setup_fn], setup_fn)); }
+    if (allowlist.count(setup_fn)) { all.reduction_stack.push_back(std::make_tuple(allowlist[setup_fn], setup_fn)); }
     else
     {
       auto base = setup_fn(name_extractor, dummy_all);
 
       if (base == nullptr)
-        all.reduction_stack.push(std::make_tuple(name_extractor.generated_name, setup_fn));
+        all.reduction_stack.push_back(std::make_tuple(name_extractor.generated_name, setup_fn));
       else
         THROW("fatal: under register_reduction() all setup functions must return nullptr");
     }
   }
+
+  // populate setup_fn -> name map to be used to lookup names in setup_base
```

Comment:
not a fan of now having two way dictionary for the same information... not sure of other options

Reply:
_setup_name_map  is not identical to reduction_stack.  _setup_name_map has two purposes.  One is to map names to setup functions.   The other is to hold onto the mapping during the recursive call to setup_base that drains the reduction stack.  By the time this information is used in the reduction setup function to pass to init_learner, it is no longer in the reduction stack..

Reply:
The state of this code is temporary, setup_base was going to evolve to use a dictionary holding string->setup_fn to enable the python reductions. This is not on master yet, but on my branch but explained this when justifying that change. I got confused and thought that was on master, so I'm wary on how that's going to look in the future (holding 2 different dicts with the same info).

Reply:
Got it.  A couple of possible solutions.  1) Encapsulate both maps in one class that provides both look ups.  2) Reduce the efficiency of one of the lookups in favor or the other look up.  We can cross that bridge at the appropriate time though.